// epicfail/dist/term.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.makeTitle = exports.Stash = void 0;
const strip_ansi_1 = __importDefault(require("strip-ansi"));
class Stash {
    constructor() {
        this.stash = [];
    }
    push(log, { extra = false } = {}) {
        this.stash.push({ body: log, extra });
    }
    toString({ extra = true } = {}) {
        return strip_ansi_1.default(this.stash
            .filter((item) => (extra ? true : !item.extra))
            .map((item) => item.body)
            .join('\n'));
    }
    toCLIString() {
        return this.stash.map((item) => item.body).join('\n');
    }
    render() {
        console.log(this.toCLIString());
    }
}
exports.Stash = Stash;
function makeTitle(color, title, ...rest) {
    return (color(`# ${title}`) + (rest.length > 0 ? ' ' + color(rest.join(' ')) : ''));
}
exports.makeTitle = makeTitle;
